File: lakesoul-spark/src/main/scala/org/apache/spark/sql/lakesoul/utils/DateTimeUtils.scala
// SPDX-FileCopyrightText: 2023 LakeSoul Contributors
//
// SPDX-License-Identifier: Apache-2.0

package org.apache.spark.sql.lakesoul.utils

import java.sql.Timestamp
import java.time._
import java.util.TimeZone
import java.util.concurrent.TimeUnit._

/**
  * Forked from [[org.apache.spark.sql.catalyst.util.DateTimeUtils]].
  * Only included the methods that are used by LakeSoulTableRel and added after Spark 2.4.
  */

/**
  * Helper functions for converting between internal and external date and time representations.
  * Dates are exposed externally as java.sql.Date and are represented internally as the number of
  * dates since the Unix epoch (1970-01-01). Timestamps are exposed externally as java.sql.Timestamp
  * and are stored internally as longs, which are capable of storing timestamps with microsecond
  * precision.
  */
object DateTimeUtils {

  // we use Int and Long internally to represent [[DateType]] and [[TimestampType]]
  type SQLDate = Int
  type SQLTimestamp = Long

  // Pre-calculated values can provide an opportunity of additional optimizations
  // to the compiler like constants propagation and folding.
  final val NANOS_PER_MICROS: Long = 1000
  final val MICROS_PER_MILLIS: Long = 1000
  final val MILLIS_PER_SECOND: Long = 1000
  final val SECONDS_PER_DAY: Long = 24 * 60 * 60
  final val MICROS_PER_SECOND: Long = MILLIS_PER_SECOND * MICROS_PER_MILLIS
  final val NANOS_PER_MILLIS: Long = NANOS_PER_MICROS * MICROS_PER_MILLIS
  final val NANOS_PER_SECOND: Long = NANOS_PER_MICROS * MICROS_PER_SECOND
  final val MICROS_PER_DAY: Long = SECONDS_PER_DAY * MICROS_PER_SECOND
  final val MILLIS_PER_MINUTE: Long = 60 * MILLIS_PER_SECOND
  final val MILLIS_PER_HOUR: Long = 60 * MILLIS_PER_MINUTE
  final val MILLIS_PER_DAY: Long = SECONDS_PER_DAY * MILLIS_PER_SECOND

  def defaultTimeZone(): TimeZone = TimeZone.getDefault()

  def getTimeZone(timeZoneId: String): TimeZone = {
    val zoneId = ZoneId.of(timeZoneId, ZoneId.SHORT_IDS)
    TimeZone.getTimeZone(zoneId)
  }

  // Converts Timestamp to string according to Hive TimestampWritable convention.
  def timestampToString(tf: TimestampFormatter, us: SQLTimestamp): String = {
    tf.format(us)
  }

  def instantToMicros(instant: Instant): Long = {
    val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND)
    val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
    result
  }

  def microsToInstant(us: Long): Instant = {
    val secs = Math.floorDiv(us, MICROS_PER_SECOND)
    val mos = Math.floorMod(us, MICROS_PER_SECOND)
    Instant.ofEpochSecond(secs, mos * NANOS_PER_MICROS)
  }

  def instantToDays(instant: Instant): Int = {
    val seconds = instant.getEpochSecond
    val days = Math.floorDiv(seconds, SECONDS_PER_DAY)
    days.toInt
  }

  /**
    * Returns the number of micros since epoch from java.sql.Timestamp.
    */
  def fromJavaTimestamp(t: Timestamp): SQLTimestamp = {
    if (t != null) {
      MILLISECONDS.toMicros(t.getTime()) + NANOSECONDS.toMicros(t.getNanos()) % NANOS_PER_MICROS
    } else {
      0L
    }
  }
}

File: native-io/lakesoul-io-java/src/main/java/com/dmetasoul/lakesoul/lakesoul/io/DateTimeUtils.java
package com.dmetasoul.lakesoul.lakesoul.io;

import java.sql.Timestamp;
import java.time.*;


public class DateTimeUtils {
    private static final int MONTHS_PER_YEAR = 12;
    private static final byte DAYS_PER_WEEK = 7;
    private static final long HOURS_PER_DAY = 24L;
    private static final long MINUTES_PER_HOUR = 60L;
    private static final long SECONDS_PER_MINUTE = 60L;
    private static final long SECONDS_PER_HOUR = 3600L;
    private static final long SECONDS_PER_DAY = 86400L;
    private static final long MILLIS_PER_SECOND = 1000L;
    private static final long MILLIS_PER_MINUTE = 60000L;
    private static final long MILLIS_PER_HOUR = 3600000L;
    private static final long MILLIS_PER_DAY = 86400000L;
    private static final long MICROS_PER_MILLIS = 1000L;
    private static final long MICROS_PER_SECOND = 1000000L;
    private static final long MICROS_PER_MINUTE = 60000000L;
    private static final long MICROS_PER_HOUR = 3600000000L;
    private static final long MICROS_PER_DAY = 86400000000L;
    private static final long NANOS_PER_MICROS = 1000L;
    private static final long NANOS_PER_MILLIS = 1000000L;
    private static final long NANOS_PER_SECOND = 1000000000L;

    // See issue SPARK-35679
    // min second cause overflow in instant to micro
    private static final long MIN_SECONDS = Math.floorDiv(Long.MIN_VALUE, MICROS_PER_SECOND);

    public static Long toMicros(Object value) {
        if (value instanceof LocalDateTime) {
            LocalDateTime d = (LocalDateTime) value;
            value = d.toInstant(ZoneOffset.UTC);
            return instantToMicros((Instant) value);
        }
        if (value instanceof Timestamp) {
            return fromTimeStamp((Timestamp) value);
        }
        if (value instanceof Instant) {
            return instantToMicros((Instant) value);
        }
        return null;
    }

    static long fromTimeStamp(Timestamp t) {
        // not optimize
        return millisToMicros(t.getTime()) + (t.getNanos() / NANOS_PER_MICROS) % MICROS_PER_MILLIS;
    }

    static long millisToMicros(long millis) {
        return Math.multiplyExact(millis, MICROS_PER_MILLIS);
    }

    static long instantToMicros(Instant i) {
//        OffsetDateTime ot = Off
//        ZonedDateTime zt = i.atZone(ZoneId.of("Asia/Shanghai"));
//        i = zt.withZoneSameInstant(ZoneId.of("UTC")).toInstant();

//        i = Instant.ofEpochSecond(1612176000);
        long epochSecond = i.getEpochSecond();
        if (epochSecond == MIN_SECONDS) {
            long us = Math.multiplyExact(epochSecond + 1, MICROS_PER_SECOND);
            return Math.addExact(us, nanoToMicros(i.getNano()) - MICROS_PER_SECOND);
        }
        long us = Math.multiplyExact(epochSecond, MICROS_PER_SECOND);
        return Math.addExact(us, nanoToMicros(i.getNano()));
    }

    static long nanoToMicros(long duration) {
        long s = 1L;
        long m = Long.MAX_VALUE / 1000L;
        if (s <= 1000L) {
            return (s == 1000L) ? duration : duration / 1000L;
        } else if (duration > m) {
            return Long.MAX_VALUE;
        } else if (duration < -m) {
            return Long.MIN_VALUE;
        } else {
            return duration * 1000L;
        }
    }
}
